rustfmt-core/rustfmt-lib/src/comment.rs

```diff
@@ -188,7 +188,7 @@ pub(crate) fn combine_strs_with_missing_comments(
     let missing_comment = rewrite_missing_comment(span, shape, context)?;
 
     if missing_comment.is_empty() {
-        if allow_extend && prev_str.len() + first_sep.len() + next_str.len() <= shape.width {
+        if allow_extend && one_line_width <= shape.width {
```

Comment:
I assume this is indeed a fix and that we don't want to use the full length of `next_str`. Otherwise `next_str` can start to exceed the shape width pretty quickly. 

For example:
```rust
fn foo() {
    match glyph {
        found @ Glyph {
            data: Gl::Com { .. },
            foo: bar,
        } => (),
        _ => (),
    };
}
```

Remains unchanged by rustfmt, but adding a few more items to the struct (or longer names) will increase the total `next_str` length and fail the conditional check resulting in a newline insertion.

```rust
fn foo() {
    match glyph {
        found @ Glyph {
            data: Gl::Com { .. },
            foo: bar,
            baz: qux,
        } => (),
        _ => (),
    };
}
```
would be reformatted to the below without this change
```rust
fn foo() {
    match glyph {
        found
        @
        Glyph {
            data: Gl::Com { .. },
            foo: bar,
            baz: qux,
        } => (),
        _ => (),
    };
}
```